Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -1119,11 +1202,165 @@ ErrorOr<u32> TagTree::read_value(u32 x, u32 y, Function<ErrorOr<bool>()> const&
 
 }
 
+static IntRect aligned_enclosing_rect(IntRect outer_rect, IntRect inner_rect, int width_increment, int height_increment)
+{
+    int new_x = (inner_rect.x() / width_increment) * width_increment;
+    int new_y = (inner_rect.y() / height_increment) * height_increment;
+    int new_right = inner_rect.width() == 0 ? new_x : ceil_div(inner_rect.right(), width_increment) * width_increment;
+    int new_bottom = inner_rect.height() == 0 ? new_y : ceil_div(inner_rect.bottom(), height_increment) * height_increment;
+    return IntRect::intersection(outer_rect, IntRect::from_two_points({ new_x, new_y }, { new_right, new_bottom }));
+}
+
+static ErrorOr<void> compute_decoding_metadata(JPEG2000LoadingContext& context)
+{
+    auto make_precinct = [&](DecodedSubBand const& sub_band, IntRect precinct_rect, int xcb_prime, int ycb_prime) -> ErrorOr<DecodedPrecinct> {
+        auto rect_covered_by_codeblocks = aligned_enclosing_rect(precinct_rect, sub_band.rect, 1 << xcb_prime, 1 << ycb_prime);
+        auto num_code_blocks_wide = rect_covered_by_codeblocks.width() / (1 << xcb_prime);
+        auto num_code_blocks_high = rect_covered_by_codeblocks.height() / (1 << ycb_prime);
```

Comment:
How big can these primes get, ie can these shifts get bigger than 32?
(also isn't `x/(1<<n) == (x>>n)` (and the aligning just a bit mask)?)

Reply:
(Just to be clear, they're not prime numbers, they're called "xcb prime" as they're slightly tweaked versions of "xck"). They're both <= 10 (see check ~8 lines into `read_coding_style_parameters`).